### PR TITLE
Add JDBC Resource Managers

### PR DESCRIPTION
**Pull request by Polber**

### CODE REVIEW

File: it/src/main/java/com/google/cloud/teleport/it/jdbc/AbstractJDBCResourceManager.java [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+  @Override
+  public String getPassword() {
+    return password;
+  }
+
+  @Override
+  public synchronized String getUri() {
+    return String.format(
+        "jdbc:%s://%s:%d/%s",
+        getDockerImageName().split(":")[0],
```
Comment (bvolpato) [inline it/src/main/java/com/google/cloud/teleport/it/jdbc/AbstractJDBCResourceManager.java:107]: Maybe we can create a method `getJdbcUrlPrefix` that can be overriden?

File: it/src/main/java/com/google/cloud/teleport/it/jdbc/AbstractJDBCResourceManager.java [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+            rowToInsert.add("'" + value + "'");
+          } else {
+            rowToInsert.add(String.valueOf(value));
+          }
+        }
+        valueList.append(String.join(",", rowToInsert)).append(")");
+
+        rowsToInsert.add(valueList.toString());
+      }
+      stmt.executeUpdate(sql.append(String.join(",", rowsToInsert)).toString());
```
Comment (bvolpato) [inline it/src/main/java/com/google/cloud/teleport/it/jdbc/AbstractJDBCResourceManager.java:222]: nit: maybe print the SQL statement being executed for better?

File: it/src/main/java/com/google/cloud/teleport/it/jdbc/AbstractJDBCResourceManager.java [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+
+      sql.append("SELECT * FROM ").append(tableName);
+      ResultSet result = stmt.executeQuery(sql.toString());
+
+      while (result.next()) {
+        int id = result.getInt(tableIds.get(tableName));
+        List<String> rowToInsert = new ArrayList<>();
+        for (int i = 2; i <= result.getMetaData().getColumnCount(); i++) {
+          rowToInsert.add(result.getString(i));
+        }
```
Comment (bvolpato) [inline it/src/main/java/com/google/cloud/teleport/it/jdbc/AbstractJDBCResourceManager.java:252]: Can we handle this better than by the index? i.e., scanning all columns and treating `tableIds.get(tableName)` differently

File: it/src/main/java/com/google/cloud/teleport/it/jdbc/AbstractJDBCResourceManager.java [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+  public synchronized List<String> getTableSchema(String tableName) {
+    String sql = "";
+    try (Connection con = driver.getConnection(getUri(), username, password)) {
+      Statement stmt = con.createStatement();
+      sql = getFirstRow(tableName);
+      ResultSet result = stmt.executeQuery(sql);
+
+      ResultSetMetaData metadata = result.getMetaData();
+      List<String> columnNames = new ArrayList<>();
+      for (int i = 1; i <= metadata.getColumnCount(); i++) {
```
Comment (bvolpato) [inline it/src/main/java/com/google/cloud/teleport/it/jdbc/AbstractJDBCResourceManager.java:275]: nit add a comment why it starts on 1

File: it/src/main/java/com/google/cloud/teleport/it/jdbc/DefaultMSSQLResourceManager.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+      this.addEnv("ACCEPT_EULA", "Y");
+      this.databaseName = databaseName;
+    }
+
+    @Override
+    public SELF withUsername(String username) {
+      if (!username.equals(AbstractJDBCResourceManager.DEFAULT_JDBC_USERNAME)) {
+        throw new UnsupportedOperationException("The username cannot be overridden to " + username);
+      }
+      return this.self();
```
Comment (bvolpato) [inline it/src/main/java/com/google/cloud/teleport/it/jdbc/DefaultMSSQLResourceManager.java:142]: Check for static instance

File: it/src/main/java/com/google/cloud/teleport/it/jdbc/JDBCResourceManager.java [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+ * the License.
+ */
+package com.google.cloud.teleport.it.jdbc;
+
+import java.sql.ResultSet;
+import java.util.List;
+import java.util.Map;
+
+/** Interface for managing JDBC resources in integration tests. */
+public interface JDBCResourceManager {
```
Comment (bvolpato) [inline it/src/main/java/com/google/cloud/teleport/it/jdbc/JDBCResourceManager.java:23]: extends ResourceManager

File: it/src/main/java/com/google/cloud/teleport/it/jdbc/JDBCResourceManagerUtils.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+  private static final int MAX_PASSWORD_LENGTH = 25;
+  private static final int MIN_TABLE_ID_LENGTH = 1;
+  private static final int MAX_TABLE_ID_LENGTH = 30;
+  private static final Pattern ILLEGAL_TABLE_CHARS = Pattern.compile("[\\/\\.]"); // i.e. [/.]
+  private static final DateTimeFormatter TIME_FORMAT =
+      DateTimeFormatter.ofPattern("yyyyMMdd_HHmmss_SSSSSS");
+
+  private static final List<Character> specialChars = Arrays.asList(
+      '‘', '~', '!', '@', '#', '$', '%', '^', '&', '*', '(', ')', '_', '\\', '-', '+', '=', '{', '}',
+      '[', ']', '/', '<', '>', ',', '.', ';', '?', ':', '|', ' ');
```
Comment (bvolpato) [inline it/src/main/java/com/google/cloud/teleport/it/jdbc/JDBCResourceManagerUtils.java:52]: nit: maybe `Arrays.asList("‘~!@#$%^&*()_\\-+={},[]/<>,.;?:|".toCharArray())`?

File: v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/PubSubToBigQuery.java [hunk 8]
```diff
@@ -258,6 +260,9 @@ public static void main(String[] args) {
 
     Options options = PipelineOptionsFactory.fromArgs(args).withValidation().as(Options.class);
     BigQueryIOUtils.validateBQStorageApiOptionsStreaming(options);
+    //    options.setWorkerDiskType(
+    //
+    // "compute.googleapis.com/projects/cloud-teleport-testing/zones/us-central1-a/diskTypes/t2a-test");
```
Comment (bvolpato) [inline v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/PubSubToBigQuery.java:265]: remove comments

File: v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/PubSubToBigQuery.java [hunk 9]
```diff
@@ -198,7 +199,8 @@ public class PubSubToBigQuery {
   public interface Options
       extends PipelineOptions,
           JavascriptTextTransformerOptions,
-          BigQueryStorageApiStreamingOptions {
+          BigQueryStorageApiStreamingOptions,
+          DataflowPipelineWorkerPoolOptions {
```
Comment (bvolpato) [inline v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/PubSubToBigQuery.java:203]: remove

File: v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/utils/JdbcConverters.java [hunk 10]
```diff
@@ -85,9 +85,18 @@ public TableRow mapRow(ResultSet resultSet) throws Exception {
                 DATE_FORMATTER.format(resultSet.getDate(i).toLocalDate()));
             break;
           case "datetime":
-            outputTableRow.set(
-                getColumnRef(metaData, i),
-                DATETIME_FORMATTER.format((TemporalAccessor) resultSet.getObject(i)));
+            Object timeObject = resultSet.getObject(i);
```
Comment (bvolpato) [inline v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/utils/JdbcConverters.java:88]: rebase

**Review by bvolpato — CHANGES_REQUESTED** [review]
Looking good, added a few minor comments, please take a look

File: v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/PubSubToBigQuery.java (hunk 9 above)
Comment (Polber) [reply v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/PubSubToBigQuery.java:203]: Done.

File: v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/PubSubToBigQuery.java (hunk 8 above)
Comment (Polber) [reply v2/googlecloud-to-googlecloud/src/main/java/com/google/cloud/teleport/v2/templates/PubSubToBigQuery.java:265]: Done.

File: it/src/main/java/com/google/cloud/teleport/it/jdbc/JDBCResourceManagerUtils.java (hunk 7 above)
Comment (Polber) [reply it/src/main/java/com/google/cloud/teleport/it/jdbc/JDBCResourceManagerUtils.java:52]: Fixed.

File: it/src/main/java/com/google/cloud/teleport/it/jdbc/JDBCResourceManager.java (hunk 6 above)
Comment (Polber) [reply it/src/main/java/com/google/cloud/teleport/it/jdbc/JDBCResourceManager.java:23]: Done.

File: it/src/main/java/com/google/cloud/teleport/it/jdbc/DefaultMSSQLResourceManager.java (hunk 5 above)
Comment (Polber) [reply it/src/main/java/com/google/cloud/teleport/it/jdbc/DefaultMSSQLResourceManager.java:142]: Done.

**Comment by pabloem** [comment]
thanks y'all for working on this. Very helpful : )

Would it be possible to add a test that runs a quick Dataflow pipeline to verify that the containerized database will have connectivity with DF pipelines?

**Comment by Polber** [comment]
Hey @pabloem, This resource manager (RM) uses the same Testcontainers framework that the other non-gcloud RM's use which have been running IT's against Dataflow on each PR, so connectivity to Dataflow for this RM is no different. I also have tested an IT against dataflow using this RM that ran fine, and will be in a subsequent PR.

File: it/src/main/java/com/google/cloud/teleport/it/jdbc/AbstractJDBCResourceManager.java (hunk 2 above)
Comment (Polber) [reply it/src/main/java/com/google/cloud/teleport/it/jdbc/AbstractJDBCResourceManager.java:222]: I added a log statement. Let me know if you think we should use a different log level or if the message should be changed.

File: it/src/main/java/com/google/cloud/teleport/it/jdbc/AbstractJDBCResourceManager.java (hunk 4 above)
Comment (Polber) [reply it/src/main/java/com/google/cloud/teleport/it/jdbc/AbstractJDBCResourceManager.java:275]: Done.

**Comment by pabloem** [comment]
cool thanks

File: it/src/main/java/com/google/cloud/teleport/it/jdbc/AbstractJDBCResourceManager.java (hunk 3 above)
Comment (Polber) [reply it/src/main/java/com/google/cloud/teleport/it/jdbc/AbstractJDBCResourceManager.java:252]: Done.

File: it/src/main/java/com/google/cloud/teleport/it/jdbc/AbstractJDBCResourceManager.java (hunk 1 above)
Comment (Polber) [reply it/src/main/java/com/google/cloud/teleport/it/jdbc/AbstractJDBCResourceManager.java]: Done.

**Review by bvolpato — APPROVED** [review]
LGTM

**Pull request merged**